edx/edx-enterprise
# ENT-94: Enroll students to a program

**Description:** This PR adds ability to enroll students into a program

**JIRA:** [ENT-94](https://openedx.atlassian.net/browse/ENT-94)

**Dependencies:** no dependencies so far

**Testing instructions:**

First, prepare the environment:
1. Create Enterprise Customer (or reuse existing)
2. Create a couple of courses (or reuse existing)
3. Launch LMS and course-discovery devstack, make sure they are working together nicely.
4. In course catalog create course runs for courses selected for this test session. Add seats to those courses (+ make sure there are courses with multiple seat types and there's at least one combination of courses that have at least one common seat type).
5. Create program(s) and include some combinations of courses into each.

Testing instructions:
1. Attempt adding single user and enrolling it into a program by UUID:
1.a. Go to course discovery admin and note program UUID (or send a GET to `course-discory/api/v1/programs` and pick one from the response)
1.b. Go to "Manage Learners" view
1.c. Fill in user email
1.d. Paste program UUID into a field
1.e. Choose one of the common enrollment modes
1.f. Submit form
1.g. Observe: user is added to corresponding list of users and (will be) enrolled into all courses in the program with the chosen enrollment mode
2. Attempt adding single user and enrolling it into a program by title - same as (1), but in step d use program title instead of ID.
3. Same as 1, but use bulk students upload - all students must be enrolled in all courses in chosen program.
4. Same as 2, but use bulk students upload.
5. Check validation:
5.a. If program was not found neither by UUID nor by name - form should display sensible error and not enroll/link student(s)
5.b. If chosen mode is not available across all selected course runs - form should display sensible error and not enroll/link student(s) and also list what common modes are available for courses in the program.
5.c. If LMS API or course discovery API is not available - it is going to crash when submitting the form (behavior not affected by this PR).
6. Make sure JS interactions work fine:
6.1. When non empty string is typed/pasted into course ID, program ID input becomes disabled and after 300 milliseconds after course modes are loaded (if user continues to type it waits) and put into course mode select.
6.2. When non empty string is typed/pasted into program ID, course ID input becomes disabled and after 300 milliseconds after default course modes are put into course mode select.
6.3. When course ID or program ID is emptied, other input becomes available immediately.

**Reviewers:**
- [x] @asadiqbal08 
- [x] @saleem-latif

CC @mattdrayer 

**Merge checklist:**

**Post merge:**

**Author concerns:**

`EnterpriseCustomerManageLearnersView` has become too big and performs too much actions: manages form, links users to EnterpriseCustomer, enrolls them into courses, etc. It makes it harder and harder to test. Might be good to start pulling it apart, or at least to split the logic of parsing user input and actually performing actions on that input (enrolling/linking students) into a separate objects.

Also, `ManageLearnersForm` starts to become too big too - might be a good place to stop and re-think the enrollment+linking approach.
